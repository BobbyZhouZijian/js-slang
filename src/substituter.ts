import * as es from 'estree'
import { parse } from './parser'
import { Context } from './types'
import * as ast from './utils/astCreator'
import { evaluateBinaryExpression } from './utils/operators'
import * as rttc from './utils/rttc'

const reducers = {
  ExpressionStatement(node: es.ExpressionStatement, context: Context) {
    const [reducedExpression] = reduce(node.expression, context)
    return [ast.expressionStatement(reducedExpression as es.Expression, node.loc!), context]
  },
  BinaryExpression(node: es.BinaryExpression, context: Context) {
    const { operator, left, right } = node
    if (left.type === 'Literal') {
      if (right.type === 'Literal') {
        const error = rttc.checkBinaryExpression(node, operator, left.value, right.value)
        if (error === undefined) {
          return [ast.literal(evaluateBinaryExpression(operator, left.value, right.value)), context]
        } else {
          throw error
        }
      } else {
        const [reducedRight] = reduce(right, context)
        const reducedExpression = ast.binaryExpression(
          operator,
          left,
          reducedRight as es.Expression,
          node.loc!
        )
        return [reducedExpression, context]
      }
    } else {
      const [reducedLeft] = reduce(node.left, context)
      const reducedExpression = ast.binaryExpression(
        operator,
        reducedLeft as es.Expression,
        right,
        node.loc!
      )
      return [reducedExpression, context]
    }
  },
  Program(node: es.Program, context: Context) {
    const [firstStatement, ...otherStatements] = node.body
    if (
      firstStatement.type === 'ExpressionStatement' &&
      firstStatement.expression.type === 'Literal'
    ) {
      return [ast.program(otherStatements, node.loc!), context]
    } else {
      const [reduced, newContext] = reduce(firstStatement, context)
      return [ast.program([reduced as es.Statement, ...otherStatements], node.loc!), newContext]
    }
  },
  /* <BAD PRACTICE>
   * for now BlockStatement is just duplicated code from Program which is BAD PRACTICE,
   * but in future BlockStatement will need to handle return as well.
   * </BAD PRACTICE>
   */
  BlockStatement(node: es.BlockStatement, context: Context) {
    const [firstStatement, ...otherStatements] = node.body
    const isFirstStatementCompletelyReduced =
      firstStatement.type === 'ExpressionStatement' && firstStatement.expression.type === 'Literal'
    if (isFirstStatementCompletelyReduced) {
      return [ast.blockStatement(otherStatements, node.loc!), context]
    } else {
      const [reduced, newContext] = reduce(firstStatement, context)
      return [
        ast.blockStatement([reduced as es.Statement, ...otherStatements], node.loc!),
        newContext
      ]
    }
  }
}

function reduce(node: es.Node, context: Context): [es.Node, Context] {
  const reducer = reducers[node.type]
  if (reducer === undefined) {
    return [node, context] // if reducer is not found we just get stuck
  } else {
    return reducer(node, context)
  }
}

export function getEvaluationSteps(code: string, context: Context): Array<[es.Node, Context]> {
  const steps: Array<[es.Node, Context]> = []
  try {
    const program = parse(code, context)
    if (program === undefined) {
      return [[parse('', context)!, context]]
    }
    steps.push([program, context])
    let [reduced, newContext] = reduce(program, context)
    while ((reduced as es.Program).body.length > 0) {
      steps.push([reduced, newContext])
      // some bug with no semis
      // tslint:disable-next-line
      ;[reduced, newContext] = reduce(reduced, newContext)
    }
    return steps
  } catch (error) {
    context.errors.push(error)
    return steps
  }
}
