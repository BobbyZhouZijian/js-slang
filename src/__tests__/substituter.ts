import { mockContext } from '../mocks/context'
import { parse } from '../parser'
import { codify, getEvaluationSteps } from '../substituter'

// source 0
test('Test basic substitution', () => {
  const code = `
    (1 + 2) * (3 + 4);
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext())
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"(1 + 2) * (3 + 4);

3 * (3 + 4);

3 * 7;

21;
"
`)
})

test('Test binary operator error', () => {
  const code = `
    (1 + 2) * ('a' + 'string');
  `
  const context = mockContext()
  const program = parse(code, context)!
  const steps = getEvaluationSteps(program, context)
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"(1 + 2) * ('a' + 'string');

3 * ('a' + 'string');

3 * \\"astring\\";
"
`)
})

test('Test two statement substitution', () => {
  const code = `
    (1 + 2) * (3 + 4);
    3 * 5;
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext(4))
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"(1 + 2) * (3 + 4);
3 * 5;

3 * (3 + 4);
3 * 5;

3 * 7;
3 * 5;

21;
3 * 5;

3 * 5;

15;
"
`)
})

test('Test unary and binary boolean operations', () => {
  const code = `
  !!!true || true;
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext())
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"!!!true || true;

!!false || true;

!true || true;

false || true;

true;
"
`)
})

test('Test ternary operator', () => {
  const code = `
  1 + -1 === 0
    ? false ? garbage : Infinity
    : anotherGarbage;
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext())
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"1 + -1 === 0 ? false ? garbage : Infinity : anotherGarbage;

0 === 0 ? false ? garbage : Infinity : anotherGarbage;

true ? false ? garbage : Infinity : anotherGarbage;

false ? garbage : Infinity;

Infinity;
"
`)
})

test('Test basic function', () => {
  const code = `
  function f(n) {
    return n;
  }
  f(5+1*6-40);
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext())
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"function f(n) {
  return n;
}
f(5 + 1 * 6 - 40);

f(5 + 1 * 6 - 40);

f(5 + 6 - 40);

f(11 - 40);

f(-29);

-29;
"
`)
})

test('Test basic bifunction', () => {
  const code = `
  function f(n, m) {
    return n * m;
  }
  f(5+1*6-40, 2-5);
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext())
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"function f(n, m) {
  return n * m;
}
f(5 + 1 * 6 - 40, 2 - 5);

f(5 + 1 * 6 - 40, 2 - 5);

f(5 + 6 - 40, 2 - 5);

f(11 - 40, 2 - 5);

f(-29, 2 - 5);

f(-29, -3);

-29 * -3;

87;
"
`)
})

test('Test "recursive" function calls', () => {
  const code = `
  function factorial(n) {
    return n === 0
      ? 1
      : n * factorial(n-1);
  }
  factorial(5);
  `
  const program = parse(code, mockContext())!
  const steps = getEvaluationSteps(program, mockContext())
  expect(steps).toMatchSnapshot()
  expect(steps.map(codify).join('\n')).toMatchInlineSnapshot(`
"function factorial(n) {
  return n === 0 ? 1 : n * factorial(n - 1);
}
factorial(5);

factorial(5);

5 === 0 ? 1 : 5 * factorial(5 - 1);

false ? 1 : 5 * factorial(5 - 1);

5 * factorial(5 - 1);

5 * factorial(4);

5 * (4 === 0 ? 1 : 4 * factorial(4 - 1));

5 * (false ? 1 : 4 * factorial(4 - 1));

5 * (4 * factorial(4 - 1));

5 * (4 * factorial(3));

5 * (4 * (3 === 0 ? 1 : 3 * factorial(3 - 1)));

5 * (4 * (false ? 1 : 3 * factorial(3 - 1)));

5 * (4 * (3 * factorial(3 - 1)));

5 * (4 * (3 * factorial(2)));

5 * (4 * (3 * (2 === 0 ? 1 : 2 * factorial(2 - 1))));

5 * (4 * (3 * (false ? 1 : 2 * factorial(2 - 1))));

5 * (4 * (3 * (2 * factorial(2 - 1))));

5 * (4 * (3 * (2 * factorial(1))));

5 * (4 * (3 * (2 * (1 === 0 ? 1 : 1 * factorial(1 - 1)))));

5 * (4 * (3 * (2 * (false ? 1 : 1 * factorial(1 - 1)))));

5 * (4 * (3 * (2 * (1 * factorial(1 - 1)))));

5 * (4 * (3 * (2 * (1 * factorial(0)))));

5 * (4 * (3 * (2 * (1 * (0 === 0 ? 1 : 0 * factorial(0 - 1))))));

5 * (4 * (3 * (2 * (1 * (true ? 1 : 0 * factorial(0 - 1))))));

5 * (4 * (3 * (2 * (1 * 1))));

5 * (4 * (3 * (2 * 1)));

5 * (4 * (3 * 2));

5 * (4 * 6);

5 * 24;

120;
"
`)
})
