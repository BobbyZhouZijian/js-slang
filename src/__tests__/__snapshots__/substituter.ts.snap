// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test basic substitution 1`] = `
Array [
  Node {
    "body": Array [
      Node {
        "end": 18,
        "expression": Node {
          "end": 17,
          "left": Node {
            "end": 6,
            "left": Node {
              "end": 2,
              "loc": SourceLocation {
                "end": Position {
                  "column": 2,
                  "line": 1,
                },
                "start": Position {
                  "column": 1,
                  "line": 1,
                },
              },
              "raw": "1",
              "start": 1,
              "type": "Literal",
              "value": 1,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 6,
                "line": 1,
              },
              "start": Position {
                "column": 1,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 6,
              "loc": SourceLocation {
                "end": Position {
                  "column": 6,
                  "line": 1,
                },
                "start": Position {
                  "column": 5,
                  "line": 1,
                },
              },
              "raw": "2",
              "start": 5,
              "type": "Literal",
              "value": 2,
            },
            "start": 1,
            "type": "BinaryExpression",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 16,
            "left": Node {
              "end": 12,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "raw": "3",
              "start": 11,
              "type": "Literal",
              "value": 3,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "line": 1,
              },
              "start": Position {
                "column": 11,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 16,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "line": 1,
                },
                "start": Position {
                  "column": 15,
                  "line": 1,
                },
              },
              "raw": "4",
              "start": 15,
              "type": "Literal",
              "value": 4,
            },
            "start": 11,
            "type": "BinaryExpression",
          },
          "start": 0,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "ExpressionStatement",
      },
    ],
    "end": 18,
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 16,
            "left": Node {
              "end": 12,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "raw": "3",
              "start": 11,
              "type": "Literal",
              "value": 3,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "line": 1,
              },
              "start": Position {
                "column": 11,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 16,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "line": 1,
                },
                "start": Position {
                  "column": 15,
                  "line": 1,
                },
              },
              "raw": "4",
              "start": 15,
              "type": "Literal",
              "value": 4,
            },
            "start": 11,
            "type": "BinaryExpression",
          },
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 7,
          },
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "loc": undefined,
          "type": "Literal",
          "value": 21,
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 18,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
]
`;

exports[`Test binary operator error 1`] = `
Array [
  Node {
    "body": Array [
      Node {
        "end": 27,
        "expression": Node {
          "end": 26,
          "left": Node {
            "end": 6,
            "left": Node {
              "end": 2,
              "loc": SourceLocation {
                "end": Position {
                  "column": 2,
                  "line": 1,
                },
                "start": Position {
                  "column": 1,
                  "line": 1,
                },
              },
              "raw": "1",
              "start": 1,
              "type": "Literal",
              "value": 1,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 6,
                "line": 1,
              },
              "start": Position {
                "column": 1,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 6,
              "loc": SourceLocation {
                "end": Position {
                  "column": 6,
                  "line": 1,
                },
                "start": Position {
                  "column": 5,
                  "line": 1,
                },
              },
              "raw": "2",
              "start": 5,
              "type": "Literal",
              "value": 2,
            },
            "start": 1,
            "type": "BinaryExpression",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 25,
            "left": Node {
              "end": 14,
              "loc": SourceLocation {
                "end": Position {
                  "column": 14,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "raw": "'a'",
              "start": 11,
              "type": "Literal",
              "value": "a",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 25,
                "line": 1,
              },
              "start": Position {
                "column": 11,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 25,
              "loc": SourceLocation {
                "end": Position {
                  "column": 25,
                  "line": 1,
                },
                "start": Position {
                  "column": 17,
                  "line": 1,
                },
              },
              "raw": "'string'",
              "start": 17,
              "type": "Literal",
              "value": "string",
            },
            "start": 11,
            "type": "BinaryExpression",
          },
          "start": 0,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 27,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "ExpressionStatement",
      },
    ],
    "end": 27,
    "loc": SourceLocation {
      "end": Position {
        "column": 27,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 25,
            "left": Node {
              "end": 14,
              "loc": SourceLocation {
                "end": Position {
                  "column": 14,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "raw": "'a'",
              "start": 11,
              "type": "Literal",
              "value": "a",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 25,
                "line": 1,
              },
              "start": Position {
                "column": 11,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 25,
              "loc": SourceLocation {
                "end": Position {
                  "column": 25,
                  "line": 1,
                },
                "start": Position {
                  "column": 17,
                  "line": 1,
                },
              },
              "raw": "'string'",
              "start": 17,
              "type": "Literal",
              "value": "string",
            },
            "start": 11,
            "type": "BinaryExpression",
          },
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 27,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 27,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Object {
            "loc": undefined,
            "type": "Literal",
            "value": "astring",
          },
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 27,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 27,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
]
`;

exports[`Test two statement substitution 1`] = `
Array [
  Node {
    "body": Array [
      Node {
        "end": 18,
        "expression": Node {
          "end": 17,
          "left": Node {
            "end": 6,
            "left": Node {
              "end": 2,
              "loc": SourceLocation {
                "end": Position {
                  "column": 2,
                  "line": 1,
                },
                "start": Position {
                  "column": 1,
                  "line": 1,
                },
              },
              "raw": "1",
              "start": 1,
              "type": "Literal",
              "value": 1,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 6,
                "line": 1,
              },
              "start": Position {
                "column": 1,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 6,
              "loc": SourceLocation {
                "end": Position {
                  "column": 6,
                  "line": 1,
                },
                "start": Position {
                  "column": 5,
                  "line": 1,
                },
              },
              "raw": "2",
              "start": 5,
              "type": "Literal",
              "value": 2,
            },
            "start": 1,
            "type": "BinaryExpression",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 16,
            "left": Node {
              "end": 12,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "raw": "3",
              "start": 11,
              "type": "Literal",
              "value": 3,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "line": 1,
              },
              "start": Position {
                "column": 11,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 16,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "line": 1,
                },
                "start": Position {
                  "column": 15,
                  "line": 1,
                },
              },
              "raw": "4",
              "start": 15,
              "type": "Literal",
              "value": 4,
            },
            "start": 11,
            "type": "BinaryExpression",
          },
          "start": 0,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "ExpressionStatement",
      },
      Node {
        "end": 25,
        "expression": Node {
          "end": 24,
          "left": Node {
            "end": 20,
            "loc": SourceLocation {
              "end": Position {
                "column": 1,
                "line": 2,
              },
              "start": Position {
                "column": 0,
                "line": 2,
              },
            },
            "raw": "3",
            "start": 19,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 5,
              "line": 2,
            },
            "start": Position {
              "column": 0,
              "line": 2,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 24,
            "loc": SourceLocation {
              "end": Position {
                "column": 5,
                "line": 2,
              },
              "start": Position {
                "column": 4,
                "line": 2,
              },
            },
            "raw": "5",
            "start": 23,
            "type": "Literal",
            "value": 5,
          },
          "start": 19,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 19,
        "type": "ExpressionStatement",
      },
    ],
    "end": 25,
    "loc": SourceLocation {
      "end": Position {
        "column": 6,
        "line": 2,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 16,
            "left": Node {
              "end": 12,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "start": Position {
                  "column": 11,
                  "line": 1,
                },
              },
              "raw": "3",
              "start": 11,
              "type": "Literal",
              "value": 3,
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "line": 1,
              },
              "start": Position {
                "column": 11,
                "line": 1,
              },
            },
            "operator": "+",
            "right": Node {
              "end": 16,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "line": 1,
                },
                "start": Position {
                  "column": 15,
                  "line": 1,
                },
              },
              "raw": "4",
              "start": 15,
              "type": "Literal",
              "value": 4,
            },
            "start": 11,
            "type": "BinaryExpression",
          },
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
      Node {
        "end": 25,
        "expression": Node {
          "end": 24,
          "left": Node {
            "end": 20,
            "loc": SourceLocation {
              "end": Position {
                "column": 1,
                "line": 2,
              },
              "start": Position {
                "column": 0,
                "line": 2,
              },
            },
            "raw": "3",
            "start": 19,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 5,
              "line": 2,
            },
            "start": Position {
              "column": 0,
              "line": 2,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 24,
            "loc": SourceLocation {
              "end": Position {
                "column": 5,
                "line": 2,
              },
              "start": Position {
                "column": 4,
                "line": 2,
              },
            },
            "raw": "5",
            "start": 23,
            "type": "Literal",
            "value": 5,
          },
          "start": 19,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 19,
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 6,
        "line": 2,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "left": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 17,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "operator": "*",
          "right": Object {
            "loc": undefined,
            "type": "Literal",
            "value": 7,
          },
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
      Node {
        "end": 25,
        "expression": Node {
          "end": 24,
          "left": Node {
            "end": 20,
            "loc": SourceLocation {
              "end": Position {
                "column": 1,
                "line": 2,
              },
              "start": Position {
                "column": 0,
                "line": 2,
              },
            },
            "raw": "3",
            "start": 19,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 5,
              "line": 2,
            },
            "start": Position {
              "column": 0,
              "line": 2,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 24,
            "loc": SourceLocation {
              "end": Position {
                "column": 5,
                "line": 2,
              },
              "start": Position {
                "column": 4,
                "line": 2,
              },
            },
            "raw": "5",
            "start": 23,
            "type": "Literal",
            "value": 5,
          },
          "start": 19,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 19,
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 6,
        "line": 2,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "loc": undefined,
          "type": "Literal",
          "value": 21,
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 18,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "type": "ExpressionStatement",
      },
      Node {
        "end": 25,
        "expression": Node {
          "end": 24,
          "left": Node {
            "end": 20,
            "loc": SourceLocation {
              "end": Position {
                "column": 1,
                "line": 2,
              },
              "start": Position {
                "column": 0,
                "line": 2,
              },
            },
            "raw": "3",
            "start": 19,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 5,
              "line": 2,
            },
            "start": Position {
              "column": 0,
              "line": 2,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 24,
            "loc": SourceLocation {
              "end": Position {
                "column": 5,
                "line": 2,
              },
              "start": Position {
                "column": 4,
                "line": 2,
              },
            },
            "raw": "5",
            "start": 23,
            "type": "Literal",
            "value": 5,
          },
          "start": 19,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 19,
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 6,
        "line": 2,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Node {
        "end": 25,
        "expression": Node {
          "end": 24,
          "left": Node {
            "end": 20,
            "loc": SourceLocation {
              "end": Position {
                "column": 1,
                "line": 2,
              },
              "start": Position {
                "column": 0,
                "line": 2,
              },
            },
            "raw": "3",
            "start": 19,
            "type": "Literal",
            "value": 3,
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 5,
              "line": 2,
            },
            "start": Position {
              "column": 0,
              "line": 2,
            },
          },
          "operator": "*",
          "right": Node {
            "end": 24,
            "loc": SourceLocation {
              "end": Position {
                "column": 5,
                "line": 2,
              },
              "start": Position {
                "column": 4,
                "line": 2,
              },
            },
            "raw": "5",
            "start": 23,
            "type": "Literal",
            "value": 5,
          },
          "start": 19,
          "type": "BinaryExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 19,
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 6,
        "line": 2,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
  Object {
    "body": Array [
      Object {
        "expression": Object {
          "loc": undefined,
          "type": "Literal",
          "value": 15,
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 2,
          },
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "type": "ExpressionStatement",
      },
    ],
    "loc": SourceLocation {
      "end": Position {
        "column": 6,
        "line": 2,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "type": "Program",
  },
]
`;
